dhall/src/Dhall/Freeze.hs

```diff
@@ -122,7 +129,28 @@ freeze inplace _standardVersion = do
             return (text, System.FilePath.takeDirectory file)
 
     (header, parsedExpression) <- parseExpr srcInfo text
-    frozenExpression <- traverse (hashImport directory _standardVersion) parsedExpression
+    frozenExpression <- traverse (tryFreeze directory) parsedExpression
     writeExpr inplace (header, frozenExpression)
         where
             srcInfo = fromMaybe "(stdin)" inplace
+
+            tryFreeze directory i = do
+                eligible <- shouldFreeze i
+                if eligible then hashImport directory _standardVersion i else return i
+
+-- | @allowPrefix x@ is an import predicate suitable for use with 'freeze' that
+-- allows any import with a common prefix.
+allowPrefix :: MonadIO io => Text -> (Import -> io Bool)
+allowPrefix prefix i =
+  case importType (importHashed i) of
+      Dhall.Core.Missing ->
+          return False
+
+      Dhall.Core.Env{} ->
+          return False
+
+      Dhall.Core.Remote (Dhall.Core.URL scheme authority file query fragment _maybeHeaders) ->
+        return (prefix `isPrefixOf` Dhall.Import.remoteToURL scheme authority file query fragment)
```

Comment:
I think it should match a logical prefix against the canonicalized path if possible.

By logical prefix I mean that if a person specifies that the prefix is `"./foo"` then it should match `"./foo/bar"` but not `"./foo.dhall"`.  Similarly, a prefix of `"."` should not match the path `"../foo"` (in fact, the prefix should give a parse error since it's an incomplete path prefix and needs to minimally be `"./"`).

The reason for canonicalizing the path before matching the prefix against it is so that a prefix of `"./"` does not match `"./../foo"`.

Reply:
But the whole point is that I want to freeze only stuff that's on `prelude.dhall-lang.org` - it seems that your suggestion would have me lose that distinction?

Reply:
@ocharles: I think it would still work because you could freeze `"https://prelude.dhall-lang.org/"` which is a valid logical prefix